src/dfx/src/lib/migrate.rs

```diff
@@ -0,0 +1,116 @@
+use anyhow::{anyhow, Context, Error};
+use candid::{CandidType, Deserialize, Principal};
+use ic_agent::Identity as _;
+use ic_utils::{
+    call::AsyncCall,
+    interfaces::{
+        management_canister::{
+            builders::{CanisterSettings, InstallMode},
+            StatusCallResult,
+        },
+        ManagementCanister, WalletCanister,
+    },
+    Argument,
+};
+
+use crate::{
+    lib::waiter::waiter_with_timeout,
+    util::{assets::wallet_wasm, expiry_duration},
+};
+
+use super::{
+    environment::Environment,
+    error::DfxResult,
+    identity::{Identity, IdentityManager},
+    models::canister_id_store::CanisterIdStore,
+    network::network_descriptor::NetworkDescriptor,
+    root_key::fetch_root_key_if_needed,
+};
+
+pub async fn migrate(env: &dyn Environment, network: &NetworkDescriptor, fix: bool) -> DfxResult {
+    fetch_root_key_if_needed(env).await?;
+    let agent = env
+        .get_agent()
+        .expect("Could not get agent from environment");
+    let mut mgr = IdentityManager::new(env)?;
+    let ident = mgr.instantiate_selected_identity()?;
+    let wallet = Identity::wallet_canister_id(env, network, ident.name())
+        .map_err(|_| anyhow!("No wallet found; nothing to do"))?;
+    let wallet = WalletCanister::create(agent, wallet)
+        .await
+        .context("No accessible wallet found; nothing to do")?;
+    let mgmt = ManagementCanister::create(agent);
+    if !wallet.version_supports_u128_cycles() {
```

Comment:
Right now the newest version is the only wallet version that supports 128-bit cycles.

After the next wallet version is released, would you expect `dfx fix` to upgrade it?

Reply:
Currently we don't tie wallet bundling into a strong version, so all we would have to go on is module hash, and there's no way of knowing whether a given module hash is newer or older. The version we *do* publish is tied to the API, and API changes would probably come with ic-utils changes, so what I've got here seems like the best option for now. Though it can be changed later if it proves not to be.